### PR TITLE
Created Inventory System

### PR DESCRIPTION
**Pull request by Christian-Irvine**
This branch creates the store inventory system and a simple GUI for it. It enables the ability to pick an item of stock/structure and it to be selected for placement. The item that has been selected is displayed in the right hand of the user as a 3D Model. The keybind to open the inventory is `Tab` and can be configured via the input system if needed in the future. Opening up the inventory will also remove any item that is currently being held from the players hand.

### CODE REVIEW

File: Assets/Scripts/InventoryItemSlot.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+using UnityEngine;
+using UnityEngine.UI;
+
+// Stores information about things on the inventory grid slots. This is so it can be easily accessed when setting it in InventoryManager
+public class InventoryItemSlot : MonoBehaviour
+{
+    [SerializeField] private Button button;
+    [SerializeField] private TextMeshProUGUI text;
+    [HideInInspector] public Button Button { get => button;  set => button = value; } 
+    [HideInInspector] public TextMeshProUGUI Text { get => text; set => text = value; }
```
Comment (lolyay270) [inline Assets/Scripts/InventoryItemSlot.cs:13]: I didn't know you could write accessors with `=>`, but I like how fluid it looks

File: Assets/Scripts/InventoryManager.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    [SerializeField] private float stockScale;
+    [SerializeField] private float structureScale;
+
+
+    private PlaceableObject heldObject;
+    public PlaceableObject HeldObject { get => heldObject; set => heldObject = value; }
+    private int tabIndex;
+    private List<GameObject> gridObjectDisplayList = new List<GameObject>();
+    //List of items currently displayed in GUI
+    private List<PlaceableObject> inventoryObjectDisplayList = new List<PlaceableObject>();
```
Comment (lolyay270) [inline Assets/Scripts/InventoryManager.cs:26]: any instance that needs `new [Type]()` can be condensed to `new()`
but that might be confusing to others

File: Assets/Scripts/InventoryManager.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    private GameObject playerHeldItem;
+    public GameObject PlayerHeldItem { get => playerHeldItem; set => playerHeldItem = value; }
+
+    // Makes InventorManager a singleton
+    private void Awake()
+    {
+        if (Instance != null)
+        {
+            Destroy(this);
+            return;
```
Comment (lolyay270) [inline Assets/Scripts/InventoryManager.cs:36]: would this not destroy the instance, then continue running the program, without making the instance again?

File: Assets/Scripts/InventoryManager.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+        int indexCount = 0;
+
+        // Iterates over all stock to see if it should display in current tab
+        inventoryPlaceableObjects.ForEach(placeableObject => {
+            if (((int)placeableObject.type) == tabIndex)
+            {
+                CreateGridItem(indexCount, placeableObject);
+
+                indexCount++;
```
Comment (lolyay270) [inline Assets/Scripts/InventoryManager.cs:91]: would this be better suited to a normal for loop, that uses `inventoryPlacableObjects.Count`,
so you don't have to make the `indexCount` variable and force it to iterate?

Or you could use a quirky foreach I found on [stackoverflow](https://stackoverflow.com/a/11437562), which implements `i` into the foreach statement

**Review by lolyay270 — APPROVED** [review]
I really like this feature, it is easy and intuitive to use.
Removing the item from the players hand when they go into the inventory is a good touch, this allows players to walk around with no objects in their hands if they want.
The placement and scale of the object in hand feels realistic for the stock items, 
and feels reasonable for the structure items to shrink to fit in the screen.
<br/>

### Ideas on where this feature could improve in the future:
image
The tabs don's have any tell to which one you're currently inside of
image
I personally really like the Windows 10 tabs that show in a file's properties window. These tabs not only show the inactive ones greyed out, but also removes the underline from the tab header to look like its connected to the page

The camera currently moves when inside the inventory menu. This is a little disorienting, but im sure its a pretty easy fix that will make a big impact.

image
Images inside of the stock item's box, that has the name attached or shows when hover. Personally I would also prefer a constant font size for the headers and item names

File: Assets/Scripts/InventoryManager.cs (hunk 4 above)
Comment (Christian-Irvine) [reply Assets/Scripts/InventoryManager.cs:91]: In this case that wouldn't quite work. The indexCount only increases if the criteria is met as the number is used assuming that is the case so a normal for loop or the special foreach would not work as intended.

Comment (lolyay270) [reply Assets/Scripts/InventoryManager.cs:91]: Ah my bad, i misread the code

**Review by lolyay270 — APPROVED** [review]
The fix you have added corrects the issue I mention prior

**Pull request merged**